Addition of PERCENT keyword in the LIMIT clause

chenxi8611 opened the pull request:
I've added percent keyword in the LIMIT clause.
@Mytherin -- please check it if it is implemented properly.

chenxi8611 committed:
Merge pull request #1 from duckdb/master

update to 2021.11.09

chenxi8611 committed:
Merge branch 'duckdb:master' into master

chenxi8611 committed:
Addition of PERCENT keyword in the LIMIT clause

chenxi8611 committed:
Merge branch 'duckdb:master' into master

Commit:
fix of #2621 for test case and compile

Mytherin commented on src/planner/binder/query_node/plan_subquery.cpp:
```diff
@@ -22,7 +22,7 @@ static unique_ptr<Expression> PlanUncorrelatedSubquery(Binder &binder, BoundSubq
 		// uncorrelated EXISTS
 		// we only care about existence, hence we push a LIMIT 1 operator
-		auto limit = make_unique<LogicalLimit>(1, 0, nullptr, nullptr);
+		auto limit = make_unique<LogicalLimit>(false, 1, 0, nullptr, nullptr);
```
Also create a `<LogicalLimitPercentage>` and leave the regular Limit as-is. That way the distinction between percentage and non-percentage based limits is more obvious and not just a boolean flag.

Mytherin commented on test/sql/order/test_limit_percent.test:
```diff
@@ -0,0 +1,109 @@
+
+# LIMIT PERCENT with order by
+query I
+SELECT i from generate_series(1, 2000) tbl(i) WHERE (i / 10 * 10 = i) ORDER BY i DESC LIMIT 10%
```
Please add the following tests:

* Tests with many values, e.g.

```sql
SELECT * FROM range(1000000) LIMIT 0.0001%
```

* Tests with LIMIT% and OFFSET combined
* LIMIT% with complex types (strings, lists, structs)
* LIMIT 0%
* Test behavior when percentage does not match exactly, e.g. if we have 10 rows, does 25% return 2 or 3 rows? What about 21%? Or 29%? That behavior should probably match Oracle/other systems that support this feature.
* LIMIT% in subqueries, e.g.:
```sql
SELECT * FROM (SELECT * FROM (SELECT * FROM range(1000000) LIMIT 10%) LIMIT 10%) LIMIT 10%;
```
* LIMIT% in set operations, e.g.:
```sql
SELECT * FROM range(10) UNION ALL SELECT * FROM range(10, 20) LIMIT 10%;
```

chenxi8611 replied:
Okay, I will add test case as you said.

Mytherin reviewed (commented):
Thanks for the PR!

I see that you can have gone for the approach of materializing everything and then selecting the top X% of the materialized result. I think that is fine and a good initial approach.

I have added some comments.

Also note that the vector_sizes test is failing. Please have a look at that. You can test that locally by changing the vector_size to 2 in `src/include/duckdb/common/vector_size.hpp`, e.g.:

```
#define STANDARD_VECTOR_SIZE 2
```

chenxi8611 commented on src/execution/operator/helper/physical_limit.cpp:
```diff
@@ -109,6 +117,47 @@ SinkResultType PhysicalLimit::Sink(ExecutionContext &context, GlobalSinkState &g
 }
 
+void PhysicalLimit::Combine(ExecutionContext &context, GlobalSinkState &gstate, LocalSinkState &lstate) const {
+	auto &state = (LimitGlobalState &)gstate;
```
Thanks for your pointers. I am now considering move it to Finalize, because it is the cause of vector_sizes test failure.

Commit:
update limit percent logic

chenxi8611 committed:
Merge branch 'duckdb:master' into master

Commit:
remove invalid comment

Commit:
add initial value

Commit:
fix format-check error

Commit:
fix warning and format error

Commit:
refix warning and format error

chenxi8611 commented:
@Mytherin I've separately created the classes for limit percent, e.g. LimitPercentModifier, BoundLimitPercentModifier, LogicalLimitPercent, LogicalTopNPercent as you said. And I've added the test cases and fixed vector_size's error.
Please review it.

Mytherin commented on src/execution/operator/helper/physical_limit.cpp:
```diff
@@ -13,14 +13,32 @@ namespace duckdb {
-		this->limit = op.limit_expression ? INVALID_INDEX : op.limit_value;
-		this->offset = op.offset_expression ? INVALID_INDEX : op.offset_value;
+		if (!op.limit_expression) {
+			this->limit = op.limit_value;
```
Please create a `PhysicalLimitPercent` that runs the limit with a percentage and revert any changes made to the physical limit. This switching on two entirely different modes of execution inside one operator is not ideal.

chenxi8611 replied:
Okay, I will add PhysicalLimitPercent.

Mytherin commented on src/common/enums/logical_operator_type.cpp:
```diff
@@ -40,8 +40,12 @@ string LogicalOperatorToString(LogicalOperatorType type) {
 	case LogicalOperatorType::LOGICAL_TOP_N:
 		return "TOP_N";
+	case LogicalOperatorType::LOGICAL_TOP_N_PERCENT:
+		return "TOP_N_PERCENT";
```
TOP_N_PERCENT is not necessary for now.

Mytherin commented on src/execution/operator/order/physical_top_n.cpp:
```diff
@@ -260,6 +268,9 @@ void TopNHeap::Finalize() {
 }
 
 void TopNHeap::Reduce() {
+	if (limit == NumericLimits<idx_t>::Maximum()) {
```
Please revert any changes made to the Top N operator. A Top-N operator that is optimized for a percentage could be created in a later PR but for now leave it as-is.

chenxi8611 replied:
Okay. I will revert any changes made to the Top N operator.
By the way, I have a question.
Then can't use at same time "order by" and "limit %" clause. isn't it?

Mytherin replied:
No, the Top N node is just an optimization. It is not required for the system to function.

Mytherin commented on src/parallel/executor.cpp:
```diff
@@ -675,4 +677,168 @@ unique_ptr<DataChunk> Executor::FetchChunk() {
 	return chunk;
 }
 
+idx_t Executor::CountChunk() {
```
Why are all these changes being made to the executor? This is out of scope for this PR and should not be necessary. Please revert.

chenxi8611 replied:
This is necessary, because we must know all source count as the result of SQL execution without "limit %" before the entire SQL execution. This code doesn't affect other parts, only affects limit percent operator.

Mytherin replied:
This approach is very unmaintainable, since the entire query execution logic is now duplicated for the sole purpose of a single operator. It does not seem efficient either, since the entire query is now run multiple times.

The limit % operator is not special, it can be implemented without any changes to the executor.

chenxi8611 replied:
Okay, I will find another way without using of executor

Mytherin commented on src/planner/binder/query_node/bind_select_node.cpp:
```diff
@@ -48,6 +48,25 @@ unique_ptr<Expression> Binder::BindDelimiter(ClientContext &context, unique_ptr<
+	ExpressionBinder expr_binder(*new_binder, context);
+	expr_binder.target_type = LogicalType::DOUBLE;
+	auto expr = expr_binder.Bind(delimiter);
+	if (expr->IsFoldable()) {
```
This is exactly the same code as `Binder::BindDelimiter<int64_t>` but with a minor change that the LogicalType is different, could you turn this into a templated function instead to avoid unnecessary code duplication?

Mytherin commented on src/planner/binder/query_node/plan_select_node.cpp:
```diff
@@ -97,21 +98,33 @@ unique_ptr<LogicalOperator> Binder::CreatePlan(BoundSelectNode &statement) {
+				PlanSubqueries(&limit_modifier.offset, &root);
+				auto limit = make_unique<LogicalLimit>(limit_modifier.limit_val, limit_modifier.offset_val,
+				                                       move(limit_modifier.limit), move(limit_modifier.offset));
+				limit->AddChild(move(root));
```
Most of the code here is repeated. Could you extract the repeated code?

chenxi8611 replied:
Okay, I will extract the repeated code

Mytherin commented on src/planner/binder/query_node/plan_select_node.cpp:
```diff
@@ -97,21 +98,33 @@ unique_ptr<LogicalOperator> Binder::CreatePlan(BoundSelectNode &statement) {
+		} else if (modifier->type == ResultModifierType::LIMIT_PERCENT_MODIFIER) {
+			auto &limit_modifier = (BoundLimitPercentModifier &)*modifier;
+			if (limit_modifier.limit) {
+				PlanSubqueries(&limit_modifier.limit, &root);
```
PlanSubqueries on the `offset` also needs to happen here.

chenxi8611 replied:
Great, it is the part I missed. I will fix it. Thanks.

Mytherin commented on test/sql/order/test_limit_percent.test:
```diff
@@ -0,0 +1,147 @@
+13
+
+query I
+SELECT a FROM test LIMIT (SELECT COUNT(*) FROM test) % OFFSET 1
```
Could you add a test that has a subquery in the offset clause?

chenxi8611 replied:
Okay, I will add a test that has a subquery in the offset clause

Mytherin reviewed (commented):
Thanks for the changes! Some comments:

Commit:
remove TopNPercent

Commit:
fix limit setting error

Commit:
fix error

Commit:
fix error

Commit:
fix limit percent clause's offset error

Commit:
fix tidy-check error

Commit:
fix format-check error

chenxi8611 committed:
Merge pull request #2 from chenxi8611/test_limit_percent

Update of LIMIT PERCENT clause(#2621)

chenxi8611 committed:
Merge pull request #3 from chenxi8611/master

Merge pull request #2 from chenxi8611/test_limit_percent

Commit:
resolve conflict files

chenxi8611 committed:
Merge pull request #7 from chenxi8611/test_limit_percent

Merge pull request #3 from chenxi8611/master

Commit:
temporary push

chenxi8611 committed:
Merge pull request #5 from chenxi8611/duckdb-master

Duckdb master

Commit:
push local reop temporarily

chenxi8611 committed:
Merge pull request #6 from chenxi8611/duckdb-master

Duckdb master

chenxi8611 committed:
Merge pull request #9 from chenxi8611/test_limit_percent

Update of LIMIT PERCENT clause

Commit:
remove all changes in client_context and executor

chenxi8611 committed:
Merge pull request #10 from chenxi8611/test_limit_percent

remove all changes in client_context and executor

Commit:
Merge branch 'master' of https://github.com/duckdb/duckdb into merge_conflict

chenxi8611 committed:
Merge pull request #11 from chenxi8611/merge_conflict

Merge conflict

chenxi8611 committed:
Merge pull request #12 from chenxi8611/test_limit_percent

merge conflict

chenxi8611 commented:
I've removed PhysicalTopNPercent and added PhysicalLimitPercent, and I've fixed some errors and added a test that has a subquery in the offset clause. Also I've changed the code so that don't use executor.
@Mytherin -- Please review it.

chenxi8611 committed:
Merge branch 'duckdb:master' into test_limit_percent

chenxi8611 committed:
Merge branch 'duckdb:master' into master

chenxi8611 committed:
Merge pull request #13 from chenxi8611/test_limit_percent

Merge branch 'duckdb:master' into test_limit_percent

chenxi8611 committed:
Merge branch 'duckdb:master' into master
